pkg/storage/tsdb/bucketindex/index.go

```diff
@@ -0,0 +1,195 @@
+package bucketindex
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+
+	cortex_tsdb "github.com/cortexproject/cortex/pkg/storage/tsdb"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	IndexFilename = "bucket-index.json"
+	IndexVersion1 = 1
+
+	SegmentsFormatUnknown       = ""
+	SegmentsFormat1Based6Digits = "1b6d"
+)
+
+// Index contains all known blocks and markers of a tenant.
+type Index struct {
+	// Version of the index format.
+	Version int `json:"version"`
+
+	// List of complete blocks (partial blocks are excluded from the index).
+	Blocks []*Block `json:"blocks"`
+
+	// List of block deletion marks.
+	BlockDeletionMarks []*BlockDeletionMark `json:"block_deletion_marks"`
```

Comment:
I explicitly called it "Block deletion mark" to avoid misunderstandings with other deletion marks we may have in the system (eg. "Tenant deletion mark").